app/assets/javascripts/discourse/app/static/prosemirror/core/plugin.js

```diff
@@ -1,15 +1,34 @@
 import { Plugin } from "prosemirror-state";
 
+/*
+  There are 3 ways to define a node view:
+
+  Setting a node view class directly (e.g. code-block)
+  `nodeViews: { nodeName: NodeViewClass }`
+
+  Setting a node view class as returned by a function, when plugin params are needed (e.g. footnote)
+  `nodeViews: { nodeName: (pluginParams) => NodeViewClass }`
+
+  Setting a node view instance as returned by a function, when plugin params are needed (e.g. image)
+  `nodeViews: { nodeName: (pluginParams) => (...args) => nodeViewInstance }`
```

Comment:
What's the guidance on _when_ you'd use each of these? Since we are only using the bottom one so far.

Reply:
The three cited examples are using each kind.

It's mostly on demand, we're supporting the three scenarios we needed so far.